mono/metadata/icall.c

```diff
@@ -2163,25 +2163,39 @@ ves_icall_MonoField_SetValueInternal (MonoReflectionFieldHandle field, MonoObjec
 		mono_gchandle_free (value_gchandle);
 }
 
+ICALL_EXPORT MonoObject*
+mono_TypedReference_ToObject (MonoTypedRef* tref);
+
 ICALL_EXPORT void
 ves_icall_System_RuntimeFieldHandle_SetValueDirect (MonoReflectionField *field, MonoReflectionType *field_type, MonoTypedRef *obj, MonoObject *value, MonoReflectionType *context_type)
 {
+	HANDLE_FUNCTION_ENTER ();
+
+	ERROR_DECL (error);
+
 	MonoClassField *f;
+	MonoObjectHandle objHandle, valueHandle;
+	MonoReflectionFieldHandle fieldHandle;
 
 	g_assert (field);
 	g_assert (obj);
 	g_assert (value);
 
 	f = field->field;
 	if (!MONO_TYPE_ISSTRUCT (m_class_get_byval_arg (f->parent))) {
-		mono_set_pending_exception (mono_get_exception_not_implemented (NULL));
-		return;
+		fieldHandle = MONO_HANDLE_NEW (MonoReflectionField, field);
+		objHandle = MONO_HANDLE_NEW (MonoObject, mono_TypedReference_ToObject (obj));
```

Comment:
This is hiding a call to `mono_error_set_pending_exception`.
Which means that we don't do any error checking before we call the next thing which may overwrite the pending exception.

It is better not to call icalls from elsewhere in the runtime, ie change `mono_TypedReference_ToObject` to call a helper, and also call that helper from here.

icalls and internal runtime functions don't always behave the same way (for example: icalls that haven't been converted to use coop handles usually call `mono_error_set_pending_exception`, which is something that we generally don't like anymore in internal runtime functions).

```c
static MonoObjectHandle
typed_reference_to_object (MonoTypedRef *tref, MonoError *error)
{
	HANDLE_FUNCTION_ENTER();
	MonoObjectHandle result;
	if (MONO_TYPE_IS_REFERENCE (tref->type)) {
		MonoObject** objp = (MonoObject **)tref->value;
		result = MONO_HANDLE_NEW (MonoObject, *objp);
	} else {
		result = mono_value_box_handle (mono_domain_get (), tref->klass, tref->value, error);
	}
    HANDLE_FUNCTION_RETURN_OBJ(result);
}

ICALL_EXPORT MonoObject*
mono_TypedReference_ToObject (MonoTypedRef* tref)
{
	ICALL_ENTRY();
	MonoObjectHandle result = typed_reference_to_object (tref, error);
	mono_error_set_pending_exception (error);
	ICALL_RETURN_OBJ (result);
	/* or convert ToObject to use handles and then its body is a one-liner */
}

...
/* in ves_icall_System_RuntimeFieldHandle_SetValueDirect */
{
	ICALL_ENTRY();
	...
	{
		objHandle = typed_reference_to_object (obj, error);
		goto_if_nok (error, leave);
		...
	}
	...

leave:
	ICALL_RETURN();
}
```